walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,126 @@
+        Collection<String> requiredAttendees = request.getAttendees(); 
+        // All attendees (required and optional) for the request meeting.
+        Collection<String> allAttendees = new ArrayList<>(requiredAttendees);
+        allAttendees.addAll(request.getOptionalAttendees());
+
+        // If the meeting duration is invalid, return an empty Collection.
+        if (meetingDuration > WHOLE_DAY.duration() || meetingDuration <= 0) {
+            return Arrays.asList();
+        }
+
+        // Find available TimeRange for both required and optional attendees.
+        Collection<TimeRange> allMeetingTimes = 
+            findMeetingTimes(events, allAttendees, meetingDuration);
+
+        // If there is meeting time for all, return allMeetingTimes so that 
+        // both required and optional can attend the meeting. 
+        if (allMeetingTimes.size() > 0) {
+            return allMeetingTimes;
+        }
+
+        // If there is no required attendee, return allMeetingTimes for 
+        // optional attendees.
+        if (requiredAttendees.isEmpty()) {
+            return allMeetingTimes;
+        }
+
+        // If not, return available TimeRange for required attendees.
+        return findMeetingTimes(
+            events, requiredAttendees, meetingDuration);
+    }
+
+    /**
+      * Find all times that the meeting could happen, given attendees and duration. 
+      * We loop through all given events to find all available TimeRanges. Pointer 
+      * rangeStart is pointing at the start of the current available TimeRange, which 
+      * is the time between the pointer and start time of the current event.If the current 
+      * available TimeRange has longer duration than the requested meeting duration, add
+      * it to available meetingTimes. Then rangeStart is updated to the end of the current
+      * event.
+      */
+    private static Collection<TimeRange> findMeetingTimes(
+        Collection<Event> events, Collection<String> attendees, long duration) {
+        Collection<TimeRange> meetingTimes = new ArrayList<TimeRange>();
+
+        // Sort the events by the start time.
+        List<Event> eventsSorted = new ArrayList<Event>(events);
+        Collections.sort(eventsSorted, Event.SORT_BY_START_ASCENDING);
+
+        int rangeStart = DAY_START; // Pointer to start of the previous event.
+
+        // Linear search the current events and add new TimeRange to 
+        // meetingTimes.
+        for (Event event : eventsSorted) {
+            Collection<String> eventAttendees = event.getAttendees();
+            TimeRange eventWhen = event.getWhen();
+            int eventStart = eventWhen.start();
+            int eventEnd = eventWhen.end();
+
+            // Check if given attendees shows up in the event.
+            if (isAttending(attendees, eventAttendees)) {
+                // Check if the current event starts before the previous event ends,
+                // or there is no gap in between.
```

Comment:
This doesn't explain why we do the check. Same for comment from lines 100-101.

Reply:
The comments are updated!